### DISCUSSION TITLE
Question about free energy function setting in one model

### DISCUSSION THREAD

**Discussion by abc-hy** [idaholab/moose#24492] [Q&A Modules: Phase field]
Hi everyone,

I have a question about setting location dependent free energy function in my model.
I have a square model and I divided it to a 3*3 grid. I want to assign different free energy function on each small square in the 3*3 grid. How to realize this?
![3fa4bfdfd5f2b192b35aed3d19d0364](https://github.com/idaholab/moose/assets/63679626/39e93755-56a3-47e1-a36e-d7d91ff57c39)

The free energy function's expression is shown as below, so for different free energy function, I need to change the 4 coefficients. How to realize this?
![image](https://github.com/idaholab/moose/assets/63679626/2c2f402f-d847-4a07-805e-1f1b3964858c)

This is how I set free energy function in my file currently,
![image](https://github.com/idaholab/moose/assets/63679626/f51f5f22-9ca5-48ec-9887-42640db8fcf3)

Thank you,
Best

**Comment by GiudGiud** [comment]
Hello

I think you'll wan to use a ParsedSubdomainGenerator to create the 3x3 grid of subdomains.
Then you can block restrict material properties. You ll create 9 DerivativeParsedMaterial, one for each subdomain

Guillaume

**Reply by abc-hy** [reply]
Since I don't want to change the other things, for example initial condition and boundary condition, I should just keep them as the same, right? I only need to create more free energy function and apply them to each block I want?

**Reply by GiudGiud** [reply]
Creating more subdomains will not affect BCs or ICs. 
You can block-restrict ICs though if you wanted.

**Reply by abc-hy** [reply]
What is block-restrict?

**Reply by GiudGiud** [reply]
setting `block=` on an object restricts it to some subdomains/block
https://mooseframework.inl.gov/source/interfaces/BlockRestrictable.html

**Reply by abc-hy** [reply]
I want to subdivide my domain, however, I always got this error message, why is this?
[df0d8_t_0d4_free_energy_blcok.8388765.txt](https://github.com/idaholab/moose/files/11578174/df0d8_t_0d4_free_energy_blcok.8388765.txt)
The following is my file:
[file.txt](https://github.com/idaholab/moose/files/11578247/file.txt)
What is the problem with my subdomain setting?

**Reply by GiudGiud** [reply]
You need order mesh generators like a tree with a single final point
```
  [./domainx]
    type = GeneratedMeshGenerator
    dim = 2
    elem_type = QUAD4
    nx = 240
    ny = 48
    nz = 0
    xmin = 0
    xmax = 4.0
    ymin = 0
    ymax = 0.8
    #zmin = 0
    #zmax = 0
    uniform_refine = 2
    #nemesis = true
    #skip_partitioning = true
  []

  [./subdomains1]
    type = ParsedSubdomainMeshGenerator
    input = domainx
    combinatorial_geometry = 'x < 1 & y > 0.4'
    block_id = 1
  []
  [./subdomains2]
    type = ParsedSubdomainMeshGenerator
    input = subdomains1
    combinatorial_geometry = 'x < 1 & y < 0.4'
    block_id = 2
  []
  [./subdomains3]
    type = ParsedSubdomainMeshGenerator
    input = subdomains2
    combinatorial_geometry = 'x > 1'
    block_id = 3
  []
```

see how there's only one exit point, subdomains3

**Reply by abc-hy** [reply]
What is only one exit point?
I changed the subdomain3 and the following is the changed file.
[file.txt](https://github.com/idaholab/moose/files/11578378/file.txt)
And I run it, but I still get that error.
[error_message.txt](https://github.com/idaholab/moose/files/11578380/error_message.txt)
Why is this?

**Reply by GiudGiud** [reply]
If you draw a graph of the mesh generators, looking at the inputs, there must be only one final one

**Reply by GiudGiud** [reply]
you can look at this page, at the gneeration / DAG part
https://mooseframework.inl.gov/syntax/Mesh/
